pkg/volume/csi/csi_plugin.go

```diff
@@ -0,0 +1,181 @@
+	csiAddrTemplate = "/tmp/kubeletproxy.sock"
+	csiTimeout      = 15 * time.Second
+)
+
+var (
+	// csiVersion supported csi version
+	csiVersion = &csipb.Version{Major: 0, Minor: 1, Patch: 0}
+)
+
+type csiPlugin struct {
+	host volume.VolumeHost
+}
+
+// ProbeVolumePlugins returns implemented plugins
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	p := &csiPlugin{
+		host: nil,
+	}
+	return []volume.VolumePlugin{p}
+}
+
+// volume.VolumePlugin methods
+var _ volume.VolumePlugin = &csiPlugin{}
+
+func (p *csiPlugin) Init(host volume.VolumeHost) error {
+	glog.Info(log("plugin initializing..."))
+	p.host = host
+	return nil
+}
+
+func (p *csiPlugin) GetPluginName() string {
+	return csiPluginName
+}
+
+func (p *csiPlugin) GetVolumeName(spec *volume.Spec) (string, error) {
+	csi, err := getCSISourceFromSpec(spec)
+	if err != nil {
+		return "", err
+	}
+	return csi.VolumeHandle, nil
+}
+
+func (p *csiPlugin) CanSupport(spec *volume.Spec) bool {
+	return spec.PersistentVolume != nil && spec.PersistentVolume.Spec.CSI != nil
+}
+
+func (p *csiPlugin) RequiresRemount() bool {
+	return false
+}
+
+func (p *csiPlugin) NewMounter(
+	spec *volume.Spec,
+	pod *api.Pod,
+	_ volume.VolumeOptions) (volume.Mounter, error) {
+	pvSource, err := getCSISourceFromSpec(spec)
+	if err != nil {
+		return nil, err
+	}
+	//addr := fmt.Sprintf(csiAddrTemplate, pvSource.Driver)
+	glog.V(4).Infof(log("setting up mounter for [volume=%v,driver=%v]", pvSource.VolumeHandle, pvSource.Driver))
+	addr := csiAddrTemplate
+	client := newCsiDriverClient("unix", addr)
+	mounter := &csiMountMgr{
+		plugin:     p,
+		spec:       spec,
+		pod:        pod,
+		podUID:     pod.UID,
+		driverName: pvSource.Driver,
+		volumeID:   pvSource.VolumeHandle,
+		csiClient:  client,
+	}
+	return mounter, nil
+}
+
+func (p *csiPlugin) NewUnmounter(specName string, podUID types.UID) (volume.Unmounter, error) {
+	glog.V(4).Infof(log("setting up unmounter for [name=%v, podUID=%v]", specName, podUID))
+	addr := csiAddrTemplate
+	client := newCsiDriverClient("unix", addr)
+	unmounter := &csiMountMgr{
+		plugin:    p,
+		podUID:    podUID,
+		csiClient: client,
+	}
+	return unmounter, nil
+}
+
+func (p *csiPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volume.Spec, error) {
+	glog.V(4).Infof(log("constructing volume spec [vol=%v, path=%v]", volumeName, mountPath))
+
+	// extract driverName/volumeId from end of mountPath
+	dir, volID := path.Split(mountPath)
+	driverName := path.Base(dir)
+
+	pv := &api.PersistentVolume{
+		ObjectMeta: meta.ObjectMeta{
+			Name: volumeName,
+		},
+		Spec: api.PersistentVolumeSpec{
+			PersistentVolumeSource: api.PersistentVolumeSource{
+				CSI: &api.CSIPersistentVolumeSource{
+					Driver:       driverName,
+					VolumeHandle: volID,
+				},
+			},
+		},
+	}
+	return volume.NewSpecFromPersistentVolume(pv, false), nil
+}
+
+func (p *csiPlugin) SupportsMountOption() bool {
+	return false
```

Comment:
CSI should be able to support this. See `VolumeCapability.MountVolume.mount_flags`. We can delay implementing this to beta if you want.

Reply:
@saad-ali Adding it as a P2 to my task list for beta - TODO P2